Content/WorldGeneration/DungeonGen/DungeonMaker.cs

```diff
@@ -0,0 +1,389 @@
+﻿using StarlightRiver.Core.Systems;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Terraria;
+using Terraria.DataStructures;
+using static StarlightRiver.Content.WorldGeneration.DungeonGen.DungeonRoom;
+
+namespace StarlightRiver.Content.WorldGeneration.DungeonGen
+{
+	public interface IRoomBuildable
+	{
+		/// <summary>
+		/// Generates a DungeonRoom instance for this builder. This should NOT place any tiles, it simply creates a positioned
+		/// template.
+		/// </summary>
+		/// <param name="x">The SECTION coordinate along the X axis of the template to place</param>
+		/// <param name="y">The SECTION coordinate along the Y axis of the template to place</param>
+		/// <returns></returns>
+		DungeonRoom MakeRoom(int x, int y);
+
+		/// <summary>
+		/// The weight to give this option when choosing between room types randomly
+		/// </summary>
+		/// <returns></returns>
+		float GetWeight();
+	}
+
+	public struct TypedRoomBuilder<T> : IRoomBuildable where T : DungeonRoom, new()
+	{
+		public readonly float weight = 1;
+
+		public TypedRoomBuilder(float weight)
+		{
+			this.weight = weight;
+		}	
+
+		public DungeonRoom MakeRoom(int x, int y)
+		{
+			var room = new T
+			{
+				topLeft = new Point16(x, y)
+			};
+			return room;
+		}
+
+		public float GetWeight()
+		{
+			return weight;
+		}
+	}
+
+	public abstract class DungeonMaker
+	{
+		/// <summary>
+		/// The amount of times to re-randomize the room template type before admitting failure to place a room
+		/// </summary>
+		public const int MAX_ROOM_RETRIES = 20;
```

Comment:
These should be virtual instead of const. Different dungeons I feel like should have different maximum retries, if they're especially lax or strict